internal/db/schema/migrations/oss/postgres/24/03_wh_network_address_dimensions.up.sql

```diff
@@ -0,0 +1,150 @@
+begin;
+
+  -- wh_upsert_network_address_dimension function writes all addresses for a
+  -- host to wh_network_address_dimension if they aren't already there, creates
+  -- a network address group which contains only the addresses for the
+  -- provided host if one doesn't exist and returns the group key which contains
+  -- exactly the set of addresses which the provided host contains.
+  create function wh_upsert_network_address_dimension(p_host_id wt_public_id)
+    returns wh_dim_key
+  as $$
+  declare
+    nag_key wh_dim_key;
+    a_key  wh_dim_key;
+  begin
+    if not exists (
+      select 1
+      from whx_network_address_dimension_source
+      where host_id = p_host_id) then
+      return 'No Addresses';
+    end if;
+
+    insert into wh_network_address_dimension (
+      address, address_type, ip_address_family, private_ip_address_indicator, dns_name, ip4_address, ip6_address
+    )
+    select
+      address, address_type, ip_address_family, private_ip_address_indicator, dns_name, ip4_address, ip6_address
+    from whx_network_address_dimension_source
+    where host_id = p_host_id
+    on conflict do nothing;
+
+    with address_list (address) as (
+        select address
+        from whx_network_address_dimension_source
+        where host_id = p_host_id
+    )
+    select distinct network_address_group_key into nag_key
+    from wh_network_address_group_membership o
+    -- At least 1 address is in the group which is also in the address list
+    where o.network_address in (select address from address_list)
+      -- The number of unique addresses in this group is the same as is in the
+      -- address_list.
+      and (select count(address) from address_list) =
+          (
+            select count(i.network_address)
+            from wh_network_address_group_membership i
+            where o.network_address_group_key = i.network_address_group_key
+          )
+      -- There are no other addresses in this group which are not in the
+      -- address_list.
+      and not exists
+      (
+            select 1
+            from wh_network_address_group_membership i
+            where o.network_address_group_key = i.network_address_group_key
+              and i.network_address not in (select address from address_list)
+      )
+    ;
+
+    if nag_key is null then
+      insert into wh_network_address_group default values returning key into nag_key;
+      for a_key in
+        select address
+        from whx_network_address_dimension_source
+        where host_id = p_host_id
+        loop
+          insert into wh_network_address_group_membership
+          (network_address_group_key, network_address)
+          values
+            (nag_key,               a_key);
+        end loop;
```

Comment:
Can this loop be replaced with a single insert statement? I tried this and the tests passed:

```sql
      insert into wh_network_address_group_membership
        (network_address_group_key, network_address)
      select nag_key, address
        from whx_network_address_dimension_source
        where host_id = p_host_id;
```

Reply:
That works great! Thanks!